instrumentation/kamon-spring/src/main/scala/kamon/instrumentation/spring/server/SpringMVCInstrumentation.scala

```diff
@@ -0,0 +1,95 @@
+import org.springframework.web.servlet.HandlerMapping
+
+import java.util.concurrent.Callable
+import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
+
+class SpringMVCInstrumentation extends InstrumentationBuilder {
+
+  /*
+   * Adds serverInstrumentation to DispatcherServlet, measures execution time
+   * of calling the doDispatch method and propagates context
+   */
+  onType("org.springframework.web.servlet.DispatcherServlet")
+    .mixin(classOf[HasServerInstrumentation.Mixin])
+    .advise(method("doDispatch"), DispatchAdvice)
+    .advise(method("render"), RenderAdvice)
+    .advise(method("processHandlerException"), ProcessHandlerAdvice)
+
+  /*
+   * Changes Callable argument of startCallableProcesing with an
+   * instrumented one that stores the context when called.
+   */
+
+  onType("org.springframework.web.context.request.async.WebAsyncManager")
+    .advise(
+      method("startCallableProcessing")
+        // is this [_] ok?
+        .and(withArgument(0, classOf[Callable[_]])), classOf[CallableWrapper])
+}
+
+object ProcessHandlerAdvice {
+  @Advice.OnMethodExit(onThrowable = classOf[Throwable], suppress = classOf[Throwable])
+  def exit(@Advice.Argument(3) throwable: Throwable): Unit =
+    Option(throwable).map(Kamon.currentSpan().fail)
+
+}
+
+object RenderAdvice {
+  @Advice.OnMethodEnter()
+  def enter(): Span =
+    Kamon.serverSpanBuilder("view.render", "spring.server").start()
+
+  @Advice.OnMethodExit()
+  def exit(@Advice.Enter span: Span): Unit =
+    span.finish()
+}
+
+object DispatchAdvice {
+  @Advice.OnMethodEnter()
+  def enter(@Advice.This dispatcherServlet: HasServerInstrumentation,
+            @Advice.Argument(0) request: HttpServletRequest,
+            @Advice.Argument(1) response: HttpServletResponse): (RequestHandler, Scope) = {
+    val requestHandler = Option(request.getAttribute("kamon-handler"))
+      .getOrElse({
+        val serverInstrumentation = dispatcherServlet.getServerInstrumentation(request)
+        val handler = serverInstrumentation.createHandler(InstrumentationUtils.requestReader(request))
+        request.setAttribute("kamon-handler", handler)
+        handler
+      }).asInstanceOf[RequestHandler] // cast because .getAttribute returns AnyRef
+    val scope = Kamon.storeContext(requestHandler.context)
+
+    // should I be tupling this like this?
+    (requestHandler, scope)
+  }
+
+  @Advice.OnMethodExit(onThrowable = classOf[Throwable], suppress = classOf[Throwable])
+  def exit(@Advice.Enter enter: (RequestHandler, Scope),
+           @Advice.Argument(0) request: HttpServletRequest,
+           @Advice.Argument(1) response: HttpServletResponse): Unit = {
+    val (handler, scope) = enter
+
+    // Extract name from path if it matches a pattern
+    // otherwise, use `http.server.request`
+    Option(request.getAttribute(HandlerMapping.BEST_MATCHING_PATTERN_ATTRIBUTE))
+      .map(_.toString)
+      .foreach(handler.span.name)
```

Comment:
Can the operation be named on the enter advice? I'm not sure if the adaptive sampler will be able to get a decent operation name.

Reply:
It cannot be named on the onEnter advice since the pattern is not available there.
The earliest that I found it was in `applyPreHandle`, so now we delay the sampling decision until we exit that method, and then name the span and take a sampling decision